test/OrchardCore.Tests/Modules/OrchardCore.Tenants/Services/TenantValidatorTests.cs

```diff
@@ -0,0 +1,124 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Localization;
+using Moq;
+using OrchardCore.Data;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Models;
+using OrchardCore.Tenants.Services;
+using OrchardCore.Tenants.ViewModels;
+using Xunit;
+
+namespace OrchardCore.Modules.Tenants.Services.Tests
+{
+    public class TenantValidatorTests
+    {
+        private readonly IList<ShellSettings> _shellSettings = new List<ShellSettings>();
+
+        public TenantValidatorTests() => SeedTenants();
+
+        [Theory]
+        [InlineData("", "tenant1", "example1.com", "Feature Profile", new[] { "The tenant name is mandatory." })]
+        [InlineData("Tenant1", "tenant1", "", "Feature", new[] { "The feature profile does not exist." })]
+        [InlineData("@Invalid Tenant", "tenant1", "example1.com", "Feature Profile", new[] { "Invalid tenant name. Must contain characters only and no spaces." })]
+        [InlineData("Tenant1", null, "  ", "Feature Profile", new[] { "Host and url prefix can not be empty at the same time." })]
+        [InlineData("Tenant1", "/tenant1", "", "Feature Profile", new[] { "The url prefix can not contain more than one segment." })]
+        [InlineData("@Invalid Tenant", "/tenant1", "", "Feature Profile", new[] { "Invalid tenant name. Must contain characters only and no spaces.", "The url prefix can not contain more than one segment." })]
```

Comment:
This is great, but you removed the positive tenant creation test. That's important too, and should remain (in such data-driven way).

Reply:
If we added them in the `TenantValidatorServices` that means we should track the creation after each call, or we should bring the `AdminControllerTests` back

One more thing, as I mentioned earlier the tenant creation is different, coz you can't control the order for execution of the inline data, that's why I used list previously

Reply:
I'm only talking about testing the validation service, i.e. add a test that tests if valid tenant configs are not mistakenly recognized as invalid. The order doesn't matter. So, exactly like this test, but a positive one.

Reply:
You didn't address this.

Reply:
I think I added a positive once, let me refer it to you